subsys/mgmt/smp_udp.c

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (c) 2019, Prevas A/S
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief UDP transport for the mcumgr SMP protocol.
+ */
+
+#include <zephyr.h>
+#include <init.h>
+#include <net/socket.h>
+#include <errno.h>
+#include <mgmt/mgmt.h>
+#include <mgmt/smp_udp.h>
+#include <mgmt/buf.h>
+#include <mgmt/smp.h>
+
+#define LOG_LEVEL CONFIG_MCUMGR_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(smp_udp);
+
+static struct zephyr_smp_transport smp_udp_transport;
+static int smp_udp_sock;
+static u8_t receive_buf[CONFIG_MCUMGR_SMP_UDP_MTU];
+static struct k_thread smp_udp_thread;
+static volatile bool stop_smp_udp;
+static K_THREAD_STACK_DEFINE(smp_udp_stack, CONFIG_MCUMGR_SMP_UDP_STACK_SIZE);
+
+static int smp_udp_tx_pkt(struct zephyr_smp_transport *zst, struct net_buf *nb)
+{
+	ARG_UNUSED(zst);
+
+	struct sockaddr *addr = net_buf_user_data(nb);
+	int ret = sendto(smp_udp_sock, nb->data, nb->len, 0,
+			 addr, sizeof(*addr));
+	mcumgr_buf_free(nb);
+
+	return ret < 0 ? MGMT_ERR_EINVAL : MGMT_ERR_EOK;
+}
+
+static uint16_t smp_udp_get_mtu(const struct net_buf *nb)
+{
+	return CONFIG_MCUMGR_SMP_UDP_MTU;
+}
+
+static int smp_udp_ud_copy(struct net_buf *dst, const struct net_buf *src)
+{
+	struct sockaddr *src_ud = net_buf_user_data(src);
+	struct sockaddr *dst_ud = net_buf_user_data(dst);
+
+	*dst_ud = *src_ud;
+
+	return MGMT_ERR_EOK;
+}
+
+static int create_socket(void)
+{
+	struct sockaddr *addr;
+
+#if defined(CONFIG_MCUMGR_SMP_UDP_IPV6)
```

Comment:
This of course works but have you considered allowing both IPv4 and IPv6 support at the same time? In other parts of Zephyr there are similar patterns where one could have either
* IPv4 only
* IPv6 only
* dual IPv4 and IPv6

support. Now the code only supports either IPv4 or IPv6 and you must know in advance how the user will access the system.

Reply:
I've been thinking a bit about this. I agree it would be nice to have both transports supported at the same time. I see a few ways of going about this:
* Single smp transport and single thread. Use select() for polling both sockets.
* Single smp transport and one thread per network transport.
* One smp transport for each network transport.

The single smp transport approaches will work fine assuming the user uses one network transport at any given moment. Otherwise, we theoretically risk mixing smp fragments from the two network transports.

Having one smp transport per network transport eliminates the risk of mixing smp fragments at the cost of some extra code.

I'm leaning towards a single smp transport to keeps things simple. Any thoughts on this?

Reply:
> leaning towards a single smp transport to keeps things simple

That sounds just fine

Reply:
Ended up adding separate smp transports for ipv4 and ipv6 as is was very little extra work.